Feat/021 mapmatching need to discuss about the exact implement of hmm

good-engineer opened the pull request:
nee

good-engineer committed:
added Road state implementation

good-engineer committed:
update map matching algorithm

good-engineer committed:
added road data

good-engineer committed:
apply .jason to .gitignore

good-engineer committed:
apply .jason to .gitignore

good-engineer committed:
edited the hmm

Commit:
implement getMapMatchingLocation

Commit:
Map display road & add bound

Commit:
implement Projection point to line in sphere(earth)

Commit:
implement getcandidate in 200m & emission probability

HoneyPot8014 commented on NaverMap/app/src/main/java/com/naver/navermap/MainActivity.kt:
```diff
@@ -120,6 +123,29 @@ class MainActivity : AppCompatActivity(), OnMapReadyCallback {
 
 
+        val assetManager: AssetManager = resources.assets
+        var inputStream: InputStream = assetManager.open("sample.json")
```
Io작업은 워커 스레드에서 이루어지는게 맞습니다.

HoneyPot8014 commented on NaverMap/app/src/main/java/com/naver/navermap/MainActivity.kt:
```diff
@@ -120,6 +123,29 @@ class MainActivity : AppCompatActivity(), OnMapReadyCallback {
         }
 
 
+        val assetManager: AssetManager = resources.assets
```
에셋파일에서 디폴트 오버레이를 그리는 부분은 함수로 분리하는게 보기 더 편할 것 같습니다

Commit:
implement transition probability by dijkstra algorithm

Commit:
probability normalization & optimization
sum probability => 100%
remove states when viterbi probability is lower than 2%

Commit:
delete class hmm

pjhh88 commented:
#30
@a111001
제가 어느정도 구현을 해보았는데 이게 맞는 구현인지 확신이 잘 안섭니다.
여기에 설명을 적어 놓았는데 잘 모르겠는 부분은 말씀해주시면 바로 추가해서 적어놓겠습니다.

Commit:
corner case(when location equals startpoint or endpoint)

good-engineer committed:
added preprocessing for location

good-engineer committed:
edited preprocessing location

Commit:
counter case
move counter variable to class field

a111001 committed:
Update sample.json

a111001 commented on NaverMap/app/src/main/java/com/naver/navermap/Viterbi.kt:
```diff
@@ -0,0 +1,396 @@
+    var currCandidates: List<RoadState>? = null
+    //roadstate, viterbi probability
+    var prevStates: List<Pair<RoadState, Double>>? = null
+    var currStates: List<Pair<RoadState, Double>>? = null
```
각각의 항목을 멤버변수로 갖고 있는 것보다, 아래와 같은 형태의 클래스로 정의하여 사용하는게 좋습니다.
```kotlin
data class TimeStep {
   var location: LatLng
   var candidates: List<Pair<RoadState, Double>>
}
```

location 정보를 입력으로 받을 때마다, TimeStep를 생성하여 관리합니다.

pjhh88 replied:
넵 감사합니다

a111001 reviewed (changes requested):
TimeStep 클래스만 추가되면, 우선 현재 작업한 내용으로 Viterbi 알고리즘을 실행하여 지도 위에 정보를 표시하면 좋을 듯합니다.

- 수신한 위치정보는 [LocationOverlay](https://navermaps.github.io/android-map-sdk/guide-ko/5-5.html)를 이용하여 표시합니다. 정확도도 함께 표현하고 싶다면, [Circle Overlay](https://navermaps.github.io/android-map-sdk/reference/com/naver/maps/map/overlay/CircleOverlay.html)를 이용합니다.
- 후보로 선택된 도로는 해당 도로의 PathOverlay의 색상을 다르게 표현하거나, 맵매칭한 좌표를 [Marker](https://navermaps.github.io/android-map-sdk/guide-ko/5-2.html)로 표현합니다.

Commit:
add data class & modify some viterbi logic

pjhh88 committed:
Merge pull request #34 from a111001/patch-1

Update sample.json

good-engineer committed:
edited mapmatching preprocessing

good-engineer committed:
merged develop to mapmatching branch

Commit:
conflict

Commit:
fix conflict

good-engineer committed:
get location every one sec and map it to road)
